### PR TITLE
Create ingress or Route for Gateway

### PR DESCRIPTION
**Pull request by rubenvp8510**
Resolves https://github.com/grafana/tempo-operator/issues/265

### CODE REVIEW

File: internal/manifests/gateway/openshift.go [hunk 1]
```diff
@@ -86,23 +86,37 @@ func clusterRoleBinding(tempo v1alpha1.TempoStack) *rbacv1.ClusterRoleBinding {
...
+	case v1alpha1.TLSRouteTerminationTypeInsecure:
+		// NOTE: insecure, no tls cfg.
+	case v1alpha1.TLSRouteTerminationTypeEdge:
+		tlsCfg = &routev1.TLSConfig{Termination: routev1.TLSTerminationEdge}
+	case v1alpha1.TLSRouteTerminationTypePassthrough:
+		tlsCfg = &routev1.TLSConfig{Termination: routev1.TLSTerminationPassthrough}
+	case v1alpha1.TLSRouteTerminationTypeReencrypt:
+		tlsCfg = &routev1.TLSConfig{Termination: routev1.TLSTerminationReencrypt}
+	default: // NOTE: if unsupported, end here.
+		return nil
```
Comment (iblancasa) [inline internal/manifests/gateway/openshift.go:101]: Should we log something in this case? Because, for me, it sounds like failing silently

**Review by iblancasa — COMMENTED** [review]
Should we add something to the E2E tests to verify this feature?

File: apis/tempo/v1alpha1/tempostack_webhook.go [hunk 2]
```diff
@@ -146,6 +146,12 @@ func (d *Defaulter) Default(ctx context.Context, obj runtime.Object) error {
 		r.Spec.Template.QueryFrontend.JaegerQuery.Ingress.Route.Termination = TLSRouteTerminationTypeEdge
 	}
 
+	// if tenant mode is Openshift, ingress type should be route by default.
+	if r.Spec.Tenants != nil && r.Spec.Tenants.Mode == OpenShift && r.Spec.Template.Gateway.Ingress.Type == IngressTypeRoute {
+		r.Spec.Template.Gateway.Ingress.Type = IngressTypeRoute
```
Comment (andreasgerstmayr) [inline apis/tempo/v1alpha1/tempostack_webhook.go:151]: afaics `r.Spec.Template.Gateway.Ingress.Type == IngressTypeRoute` should be removed from the condition

File: internal/manifests/gateway/openshift.go (hunk 1 above)
Comment (andreasgerstmayr) [reply internal/manifests/gateway/openshift.go:101]: `TLSRouteTerminationType` is an enum in the CRD, so this case should never happen in theory.
Unless we update the list of valid enum values and forget to update this switch, so I guess yep, it makes sense to print a log message (could you add the same log message to the `switch` in the query_frontend also?).

File: apis/tempo/v1alpha1/tempostack_webhook.go [hunk 3]
```diff
@@ -146,6 +146,12 @@ func (d *Defaulter) Default(ctx context.Context, obj runtime.Object) error {
 		r.Spec.Template.QueryFrontend.JaegerQuery.Ingress.Route.Termination = TLSRouteTerminationTypeEdge
 	}
 
+	// if tenant mode is Openshift, ingress type should be route by default.
```
Comment (pavolloffay) [inline apis/tempo/v1alpha1/tempostack_webhook.go:149]: @rubenvp8510 should there be a validation for jquery ingeress and gateway ingress? For instance do not allow Jquery ingress if the gateway is deployed?

**Review by pavolloffay — COMMENTED** [review]
LGTM just a minor comment

File: internal/manifests/gateway/openshift.go (hunk 1 above)
Comment (pavolloffay) [reply internal/manifests/gateway/openshift.go:101]: It should be checked by link https://github.com/grafana/tempo-operator/blob/7c7ee6470cd01138f1225d8734dd7f3e7229135f/.golangci.yaml#L29

**Review by iblancasa — APPROVED** [review]

File: apis/tempo/v1alpha1/tempostack_webhook.go (hunk 3 above)
Comment (rubenvp8510) [reply apis/tempo/v1alpha1/tempostack_webhook.go:149]: Yes, that is true, we need to validate it.

Comment (rubenvp8510) [reply apis/tempo/v1alpha1/tempostack_webhook.go:150]: Indeed I think that validation already happens here: https://github.com/grafana/tempo-operator/blob/main/apis/tempo/v1alpha1/tempostack_webhook.go#L254

**Review by pavolloffay — APPROVED** [review]

**Pull request merged**